## Exercises

1. Load and display the image 'dog.jfif'. Save the image in png format.
2. *Suggest 2 ways and write codes to display 2 images simultaneously.* You can use any image snapped from your handphone, downloaded from internet or images from weekly materials on MS teams. The 2 images are original color image and its corresponding grayscale image.
3. Write codes that performs the following:
    * Load the video “img_pexels.mp4” into the Python environment, resize it and display the videos with smaller frames (The frames can be of any size, as long as it is smaller). You can specify an arbitrary frame rate.
    * Save it as a separate files: “smaller_img_pexels.avi” or "smaller_img_pexels.mp4"
4. Enlarge the image "dog.jfif" by using different techniques:
    1) Linear interpolation
   2) Cubic interpolation
   3) Nearest neighbor interpolation.

Perform profiling on each method. Comment on the **execution times** and **quality of resulting images**.

In [1]:
import cv2 as cv
import numpy as np
import sys
assert sys.version_info >=(3,8)

## Solution 1

In [28]:
# this assumes that the user has an images folder in working directory
img = cv.imread("images/dog.jfif")

if img is None:
    sys.exit("The file path is not detected")

cv.imshow("dog",img)
k = cv.waitKey(0) & 0xFF
if k == ord("s"):
    cv.imwrite("dog.png",img)
cv.destroyAllWindows()

## Solution 2 : Method 1

In [3]:
img = cv.imread("images/soccer.jpg")  #color image
img_grayscale = cv.imread("images/soccer.jpg",0)  #to grayscale

cv.imshow("color",img)  #first parameter is label, second parameter is variable name
cv.imshow("grayscale",img_grayscale)
cv.waitKey(0)
cv.destroyAllWindows()

## Solution 2 : Method 2

Supposingly this method's purpose is to:
- Fit 2 images in the same window
- We could maybe do it horizontally or vertically depending on our requirements

In [7]:
#Check size of color image and grayscale image
img = cv.imread("images/soccer.jpg")
img_grayscale = cv.imread("images/soccer.jpg",0)

print("Shape of img " , img.shape)
print("Shape of img_grayscale " , img_grayscale.shape)

#since they both have different dimensions
#we need to add an extra dimension to grayscale image

Shape of img  (1200, 800, 3)
Shape of img_grayscale  (1200, 800)


In [11]:
pip install scikit-image

     --------------------------------------- 12.9/12.9 MB 28.4 MB/s eta 0:00:00
     ---------------------------------------- 1.7/1.7 MB 54.6 MB/s eta 0:00:00
     ---------------------------------------- 225.9/225.9 kB ? eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


In [22]:
#Reference : 
#[1] https://stackoverflow.com/questions/59219210/extend-a-greyscale-image-to-fit-a-rgb-image
#[2] https://stackoverflow.com/questions/42920201/how-to-combine-a-rgb-image-with-a-grayed-image-in-opencv
#[3] https://stackoverflow.com/questions/7589012/combining-two-images-with-opencv/24522170
#[4] https://stackoverflow.com/questions/57577021/converting-a-2d-array-grayscale-image-into-3d-rgb-image-array

#the below code mainly uses the idea of NumPy and Scikit-image libraries
import numpy as np
from skimage import io

img = cv.imread("images/soccer.jpg")
img_grayscale = cv.imread("images/soccer.jpg",0)

#this won't work, they are different dimensions
# numpy_horizontal_concat = np.concatenate((img,img_grayscale),axis=1)

rows_rgb, cols_rgb,channels = img.shape
rows_gray, cols_gray = img_grayscale.shape

rows_comb = max(rows_rgb,rows_gray)
cols_comb = cols_rgb + cols_gray
comb = np.zeros(shape=(rows_comb,cols_comb,channels),dtype=np.uint8)

comb[:rows_rgb, :cols_rgb] = img
comb[:rows_gray, cols_rgb:] = img_grayscale[:,:,None]

cv.imshow("Side by side " , comb)
cv.waitKey(0)
cv.destroyAllWindows()
                          

## Solution 2 : Method 3

In [27]:
import cv2 as cv
import numpy as np

# Read the original and grayscale images
img = cv.imread("images/soccer.jpg")
img_grayscale = cv.imread("images/soccer.jpg", 0)

# Create an RGB version of the grayscale image by stacking the grayscale image three times
# we are basically stacking Shape of img_grayscale  (1200, 800) 3 times
#because originally we only had a 2d array
# we want to convert 2d grayscale image into 3d rgb image

RGB_img_grayscale = np.dstack((img_grayscale, img_grayscale, img_grayscale))

# Concatenate the original and the new RGB grayscale image side by side
numpy_horizontal_concat = np.concatenate((img, RGB_img_grayscale), axis=1)

print(RGB_img_grayscale.shape)
# Display the concatenated image
cv.imshow("Side by side", numpy_horizontal_concat)
cv.waitKey(0)
cv.destroyAllWindows()


(1200, 800, 3)


## Useful reading to understand RGB better
- https://medium-parser.vercel.app/?url=https://towardsdatascience.com/exploring-the-mnist-digits-dataset-7ff62631766a
- https://datahacker.rs/convolution-rgb-image/

## Solution 3

In [65]:
#in order to save a video file, must create a video write object

cap = cv.VideoCapture("videos/img_pexels.mp4")

if not cap.isOpened():
    raise Exception("No video file detected")

width = int(cap.get(3))  #default dimension
height = int(cap.get(4)) #default dimension
fourcc = cv.VideoWriter_fourcc("M","J","P","G")   

new_width = 300
new_height = 150

fps = 60
out = cv.VideoWriter("smaller_img_pexels.avi",fourcc,fps,(new_width,new_height))  #used to write video frames to a file

while cap.isOpened():
    ret,frame = cap.read()

    if not ret:
        print("Can't receive frame")
        break

    resize_frame = cv.resize(frame, (new_width,new_height)) 
    out.write(resize_frame)  #writes current frame to output video file
    
    cv.imshow("resized video", resize_frame)
    if cv.waitKey(1) & 0xFF == 27:
        break
        

cap.release()
out.release()  #closes the video file
cv.destroyAllWindows()

## Solution 4 : Method 1

In [5]:
pip install line_profiler

     ------------------------------------ 125.5/125.5 kB 922.3 kB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


In [6]:
%load_ext line_profiler

In [39]:
import cv2 as cv
def display_images(images,titles):
    for image, title in zip (images,titles):
        cv.imshow(title,image)

    cv.waitKey(0)
    cv.destroyAllWindows()

In [47]:
#Linear interpolation

def method_1() :
    img = cv.imread("images/dog.jfif")
    new_dim = (1200,800)
    img_resize = cv.resize(img,new_dim,interpolation=cv.INTER_LINEAR)
    display_images([img,img_resize],("original","resize"))

%lprun -f method_1 method_1()

Timer unit: 1e-07 s

Total time: 10.0327 s

Could not find file C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\3532763709.py
Are you sure you are running this program from the same directory
that you ran the profiler from?
Continuing without the function's contents.

Line #      Hits         Time  Per Hit   % Time  Line Contents
     3                                           
     4         1      13080.0  13080.0      0.0  
     5         1         14.0     14.0      0.0  
     6         1      11699.0  11699.0      0.0  
     7         1  100301876.0    1e+08    100.0

## Solution 4 : Method 2

In [49]:
#Cubic interpolation
def method_2() :
    img = cv.imread("images/dog.jfif")
    new_dim = (1200,800)
    img_resize = cv.resize(img,new_dim,interpolation=cv.INTER_CUBIC)
    display_images([img,img_resize],("original","resize"))

%lprun -f method_1 method_2()

Timer unit: 1e-07 s

Total time: 3.03032 s

Could not find file C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\3532763709.py
Are you sure you are running this program from the same directory
that you ran the profiler from?
Continuing without the function's contents.

Line #      Hits         Time  Per Hit   % Time  Line Contents
     3                                           
     4         1         14.0     14.0      0.0  
     5         1      13999.0  13999.0      0.0  
     6         1   30289202.0    3e+07    100.0

## Solution 4 : Method 3

In [48]:
#Nearest Neighbor interpolation
def method_3() :
    img = cv.imread("images/dog.jfif")
    new_dim = (1200,800)
    img_resize = cv.resize(img,new_dim,interpolation=cv.INTER_NEAREST)
    display_images([img,img_resize],("original","resize"))

%lprun -f method_1 method_3()

Timer unit: 1e-07 s

Total time: 2.10187 s

Could not find file C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\3532763709.py
Are you sure you are running this program from the same directory
that you ran the profiler from?
Continuing without the function's contents.

Line #      Hits         Time  Per Hit   % Time  Line Contents
     3                                           
     4         1        221.0    221.0      0.0  
     5         1      12072.0  12072.0      0.1  
     6         1   21006390.0    2e+07     99.9

## Using Cprofile

In [27]:
import cProfile
import pstats

profiler = cProfile.Profile()
profiler.enable()

method_1()

profiler.disable()
stats = pstats.Stats(profiler).sort_stats('cumtime')
stats.print_stats()


         45 function calls in 0.128 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000    0.128    0.064 C:\Users\kok shen\AppData\Local\Programs\Python\Python310\lib\site-packages\IPython\core\interactiveshell.py:3512(run_code)
        2    0.000    0.000    0.128    0.064 {built-in method builtins.exec}
        1    0.000    0.000    0.128    0.128 C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\1633686941.py:10(method_1)
        1    0.000    0.000    0.126    0.126 C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\1633686941.py:3(display_images)
        2    0.094    0.047    0.094    0.047 {imshow}
        1    0.033    0.033    0.033    0.033 {destroyAllWindows}
        1    0.001    0.001    0.001    0.001 {imread}
        1    0.000    0.000    0.000    0.000 {resize}
        2    0.000    0.000    0.000    0.000 C:\Users\kok shen\AppData\Local\Programs\Python\Python310\lib\codeop.

In [28]:
import cProfile
import pstats

profiler = cProfile.Profile()
profiler.enable()

method_2()

profiler.disable()
stats = pstats.Stats(profiler).sort_stats('cumtime')
stats.print_stats()

         45 function calls in 0.207 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000    0.207    0.103 C:\Users\kok shen\AppData\Local\Programs\Python\Python310\lib\site-packages\IPython\core\interactiveshell.py:3512(run_code)
        2    0.000    0.000    0.207    0.103 {built-in method builtins.exec}
        1    0.000    0.000    0.207    0.207 C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\3666958759.py:2(method_2)
        1    0.000    0.000    0.205    0.205 C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\1633686941.py:3(display_images)
        2    0.172    0.086    0.172    0.086 {imshow}
        1    0.033    0.033    0.033    0.033 {destroyAllWindows}
        1    0.001    0.001    0.001    0.001 {imread}
        1    0.000    0.000    0.000    0.000 {resize}
        2    0.000    0.000    0.000    0.000 C:\Users\kok shen\AppData\Local\Programs\Python\Python310\lib\codeop.p

In [29]:
import cProfile
import pstats

profiler = cProfile.Profile()
profiler.enable()

method_3()

profiler.disable()
stats = pstats.Stats(profiler).sort_stats('cumtime')
stats.print_stats()

         45 function calls in 0.278 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000    0.278    0.139 C:\Users\kok shen\AppData\Local\Programs\Python\Python310\lib\site-packages\IPython\core\interactiveshell.py:3512(run_code)
        2    0.000    0.000    0.278    0.139 {built-in method builtins.exec}
        1    0.000    0.000    0.278    0.278 C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\2918942059.py:1(<module>)
        1    0.000    0.000    0.278    0.278 C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\313599484.py:2(method_3)
        1    0.000    0.000    0.276    0.276 C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\1633686941.py:3(display_images)
        2    0.217    0.108    0.217    0.108 {imshow}
        1    0.059    0.059    0.059    0.059 {destroyAllWindows}
        1    0.001    0.001    0.001    0.001 {imread}
        1    0.000    0.000    0.000    0.000 {

## Final Solution : Combining profiling methods together

In [37]:
import cv2 as cv
import cProfile
import pstats
from io import StringIO

def display_images(images,titles):
    for image, title in zip (images,titles):
        cv.imshow(title,image)

    # cv.waitKey(0)
    cv.destroyAllWindows()

#Linear interpolation
def method_1() :
    img = cv.imread("images/dog.jfif")
    new_dim = (300,200)
    img_resize = cv.resize(img,new_dim,interpolation=cv.INTER_LINEAR)
    display_images([img,img_resize],("original","resize"))

#Cubic interpolation
def method_2() :
    img = cv.imread("images/dog.jfif")
    new_dim = (300,200)
    img_resize = cv.resize(img,new_dim,interpolation=cv.INTER_CUBIC)
    display_images([img,img_resize],("original","resize"))

#Nearest Neighbor interpolation
def method_3() :
    img = cv.imread("images/dog.jfif")
    new_dim = (300,200)
    img_resize = cv.resize(img,new_dim,interpolation=cv.INTER_NEAREST)
    display_images([img,img_resize],("original","resize"))



def profile_method(method, method_name):
    profiler = cProfile.Profile()
    profiler.enable()
    
    method()
    
    profiler.disable()
    s = StringIO()
    stats = pstats.Stats(profiler, stream=s).sort_stats('cumtime')
    stats.print_stats()
    
    print(f"Profiling results for {method_name}:")
    print(s.getvalue())

profile_method(method_1, "Linear Interpolation")
profile_method(method_2, "Cubic Interpolation")
profile_method(method_3, "Nearest Neighbor interpolation")


Profiling results for Linear Interpolation:
         8 function calls in 0.293 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.293    0.293 C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\2003948927.py:14(method_1)
        1    0.000    0.000    0.292    0.292 C:\Users\kok shen\AppData\Local\Temp\ipykernel_20944\2003948927.py:6(display_images)
        2    0.190    0.095    0.190    0.095 {imshow}
        1    0.102    0.102    0.102    0.102 {destroyAllWindows}
        1    0.001    0.001    0.001    0.001 {imread}
        1    0.000    0.000    0.000    0.000 {resize}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}



Profiling results for Cubic Interpolation:
         8 function calls in 0.113 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    

## It can be shown that :
1. Nearest Neighbor interpolation is the fastest (0.089 seconds)
2. Cubic interpolation is the second fastest (0.113 seconds)
3. Linear interpolation is the slowest (0.293 seconds)

>Additional Reference : <br>
[1] https://stackoverflow.com/questions/3112364/how-do-i-choose-an-image-interpolation-method-emgu-opencv
<br>[2] https://medium.com/htx-s-s-coe/image-super-resolution-a-comparison-between-interpolation-deep-learning-based-techniques-to-25e7531ab207

- As for quality of images :
  - Nearest neighbour :Results in an image with higher resolution, but will look blocky and unnatural
  - Cubic : Results in sharper image compared to Nearest neighbor and Linear
  - Linear : Results in smoother results than nearest neighbor, but edges don't look good